### PR TITLE
Convert composite evaluators to async based implementation

### PR DESCRIPTION
**Pull request by ninghu**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/microsoft/promptflow/blob/main/CONTRIBUTING.md).**
- [ ] **I confirm that all new dependencies are compatible with the MIT license.**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by nagkumar91 — APPROVED** [review]

File: src/promptflow-evals/promptflow/evals/evaluators/_chat/_chat.py [hunk 1]
```diff
@@ -287,3 +235,75 @@ def _validate_conversation(self, conversation: List[Dict]):
...
         if expected_role != "user":
             raise ValueError("The conversation must end with an assistant's turn.")
+
+
+class ChatEvaluator:
+    """
+    Initialize a chat evaluator configured for a specific Azure OpenAI model.
+
+    :param model_config: Configuration for the Azure OpenAI model.
+    :type model_config: AzureOpenAIModelConfiguration
```
Comment (diondrapeck) [inline src/promptflow-evals/promptflow/evals/evaluators/_chat/_chat.py:245]: nit: for reference documentation, we want to use the full path for any internal objects.

i.e. `~promptflow.core.AzureOpenAIModelConfiguration`

Full documentation guide [here](https://github.com/microsoft/promptflow/blob/main/docs/dev/documentation_guidelines.md)

File: src/promptflow-evals/promptflow/evals/evaluators/_chat/_chat.py [hunk 2]
```diff
@@ -287,3 +235,75 @@ def _validate_conversation(self, conversation: List[Dict]):
...
+    :param eval_last_turn: Set to True to evaluate only the most recent exchange in the dialogue,
+        focusing on the latest user inquiry and the assistant's corresponding response. Defaults to False
+    :type eval_last_turn: bool
+    :param parallel: If True, use parallel execution for evaluators. Else, use sequential execution.
+        Default is True.
+    :type parallel: bool
+    :return: A function that evaluates and generates metrics for "chat" scenario.
+    :rtype: function
```
Comment (diondrapeck) [inline src/promptflow-evals/promptflow/evals/evaluators/_chat/_chat.py:253]: nit: function type in Python is "Callable"

File: src/promptflow-evals/promptflow/evals/evaluators/_chat/retrieval/_retrieval.py [hunk 3]
```diff
@@ -122,3 +89,58 @@ def __call__(self, *, conversation, **kwargs):
...
             },
         }
+
+
+class RetrievalChatEvaluator:
+    """
+    Initialize an evaluator configured for a specific Azure OpenAI model.
+
+    :param model_config: Configuration for the Azure OpenAI model.
+    :type model_config: AzureOpenAIModelConfiguration
```
Comment (diondrapeck) [inline src/promptflow-evals/promptflow/evals/evaluators/_chat/retrieval/_retrieval.py:99]: same nit as above

File: src/promptflow-evals/promptflow/evals/evaluators/_content_safety/_content_safety_chat.py [hunk 4]
```diff
@@ -247,3 +181,80 @@ def _get_harm_severity_level(self, harm_score: float) -> str:
...
+        It contains subscription id, resource group, and project name.
+    :type project_scope: dict
+    :param eval_last_turn: Set to True to evaluate only the most recent exchange in the dialogue,
+        focusing on the latest user inquiry and the assistant's corresponding response. Defaults to False
+    :type eval_last_turn: bool
+    :param parallel: If True, use parallel execution for evaluators. Else, use sequential execution.
+        Default is True.
+    :type parallel: bool
+    :param credential: The credential for connecting to Azure AI project.
+    :type credential: TokenCredential
```
Comment (diondrapeck) [inline src/promptflow-evals/promptflow/evals/evaluators/_content_safety/_content_safety_chat.py:200]: ~azure.core.credentials.TokenCredential

File: src/promptflow-evals/promptflow/evals/evaluators/_content_safety/common/evaluate_with_rai_service.py [hunk 5]
```diff
@@ -29,6 +29,24 @@
 USER_AGENT = "{}/{}".format("promptflow-evals", version)
 
 
+def get_common_headers(token: str) -> Dict:
```
Comment (diondrapeck) [inline src/promptflow-evals/promptflow/evals/evaluators/_content_safety/common/evaluate_with_rai_service.py:32]: nice helper method

**Review by diondrapeck — CHANGES_REQUESTED** [review]

File: src/promptflow-evals/promptflow/evals/evaluators/_chat/_chat.py (hunk 1 above)
Comment (ninghu) [reply src/promptflow-evals/promptflow/evals/evaluators/_chat/_chat.py:245]: Got it. Will address all those documentation comments in a follow up PR.

**Review by nick863 — APPROVED** [review]

File: src/promptflow-evals/promptflow/evals/evaluators/_content_safety/common/evaluate_with_rai_service.py [hunk 6]
```diff
@@ -40,15 +58,11 @@ async def ensure_service_availability(rai_svc_url: str, token: str, capability:
...
-        "Content-Type": "application/json",
-        "User-Agent": USER_AGENT,
-    }
-
+    headers = get_common_headers(token)
     svc_liveness_url = rai_svc_url + "/checkannotation"
+
     async with httpx.AsyncClient() as client:
-        response = await client.get(svc_liveness_url, headers=headers)
+        response = await client.get(svc_liveness_url, headers=headers, timeout=60)
```
Comment (MilesHolland) [inline src/promptflow-evals/promptflow/evals/evaluators/_content_safety/common/evaluate_with_rai_service.py:65]: It seems a bit odd to have a hardcoded timeout like this. Should we make a follow-up task to replace all timeouts with a constant and/or environment variable?

Comment (ninghu) [reply src/promptflow-evals/promptflow/evals/evaluators/_content_safety/common/evaluate_with_rai_service.py:65]: Sure, will do!

File: src/promptflow-evals/promptflow/evals/evaluators/_fluency/_fluency.py [hunk 7]
```diff
@@ -39,7 +43,7 @@ async def __call__(self, *, question: str, answer: str, **kwargs):
             raise ValueError("Both 'question' and 'answer' must be non-empty strings.")
 
         # Run the evaluation flow
-        llm_output = await self._flow(question=question, answer=answer)
+        llm_output = await self._flow(question=question, answer=answer, timeout=600, **kwargs)
```
Comment (MilesHolland) [inline src/promptflow-evals/promptflow/evals/evaluators/_fluency/_fluency.py:46]: Why is the timeout 10X here compared to the content safety time?

**Review by MilesHolland — APPROVED** [review]

**Review by diondrapeck — APPROVED** [review]

**Pull request merged**